### ISSUE TITLE
spec changes for header and mirror based routing

### ISSUE THREAD

**Issue by zachaller** [argoproj/argo-rollouts#2073]
# Summary

We are planning to add traffic mirror and header based routing which now puts a different set of requirements on Argo Rollouts to have the idea of precedence within the routes that argo rollouts creates. Most traffic routers such as Istio, Ambassador, and Trafiek and others have the concept of being able to control route precedence. So in order to implement the two mentioned features above I propose adding a new managedRoutes section to traffic routing that defines the route precedence as well as allows us to clean up our routes when we are done with the rollout.

```
      trafficRouting:
        managedRoutes: # The order of the routes here become the precedence in the upstream traffic manager
          - name: set-header-1
          - name: mirror-route-3
```

```
      steps:
        - setMirrorRoute:
            name: mirror-route-3
            percentage: 100
            match:
              - method:
                  exact: POST
                path:
                  prefix: /
        - setHeaderRoute:
            name: set-header-1
            match:
            - headerName: agent-1a
              headerValue:
                regex: firefox1(.*)
```

The above spec changes are at the very minimum required for the two mentioned features. However, in order to also support a more gitops friendly workflow I think changing the managed routes to also have the option to include the route that acts as the canary route a very useful feature so that we do not have to predefine our routes such as required with istio.  Notice the `canaryRoute: true`

```
      trafficRouting:
        managedRoutes:
          - name: set-header-1
          - name: mirror-route-3
          - name: primary
            canaryRoute: true #This becomes the default/canary route in the upstream traffic router
```

Argo rollouts will then be able to place the routes that rollouts creates at a higher precedence than any user defined routes which allows the user to create their routes as if Argo Rollouts did not exists and we would then only create higher precedence routes as needed during the rollout. This allow for a very gitops friendly deployment of argo rollouts with argocd.

**Comment by zachaller** [comment]
closed by #2074

**Comment by zachaller** [comment]
https://cloud-native.slack.com/archives/C01U781DW2E/p1666632541542379?thread_ts=1666277093.264039&cid=C01U781DW2E

User complaint about not having canaryTemplate
